### PR TITLE
fix: add timeout for executor signal

### LINKED ISSUE #13011: add timeout for executor signal
### Pre-requisites

- [X] I have double-checked my configuration
- [X] I have tested with the `:latest` image tag (i.e. `quay.io/argoproj/workflow-controller:latest`) and can confirm the issue still exists on `:latest`. If not, I have explained why, **in detail**, in my description below.
- [X] I have searched existing issues and could not find a match for this bug
- [X] I'd like to contribute the fix myself (see [contributing guide](https://github.com/argoproj/argo-workflows/blob/main/docs/CONTRIBUTING.md))

### What happened/what did you expect to happen?

The method ExecPodContainerAndGetOutput in signal.go will stuck, ince this method  (GetExecutorOutput) blocks while waiting for the remote stdout output, causing the goroutine to block for a long time, or even for a long time, we should add a timeout to prevent this problem from happening again.

This causes the cleanup to block all four goroutines by default, thus growing indefinitely in the pod_clean_up queue depth

### Version

main

### Paste a small workflow that reproduces the issue. We must be able to run the workflow; don't enter a workflows that uses private images.

```YAML
none
```

### Logs from the workflow controller

```text
none
```

### Logs from in your workflow's wait container

```text
none
```

### PR DESCRIPTION
**Pull request by fyp711**
Fixes #13011

### Modifications

I added a context timeout to prevent spdy from blocking for a long time

### CODE REVIEW

**Comment by fyp711** [comment]
Could someone take a discussion for this issue ? It will block all pod cleanup workers. It's very serious.

File: workflow/signal/signal.go [hunk 1]
```diff
@@ -15,6 +17,8 @@ import (
 	"github.com/argoproj/argo-workflows/v3/workflow/common"
 )
 
+var spdyTimeout = env.LookupEnvDurationOr("SPDY_TIMEOUT", 10*time.Minute)
```
Comment (Joibel) [inline workflow/signal/signal.go:20]: The name of the environment variable to control this is difficult to 'discover'.

I'd suggest `POD_OUTPUT_TIMEOUT` or something along those lines. No user cares about the protocol in use.

This will also need adding to the documentation.

File: workflow/signal/signal.go [hunk 2]
```diff
@@ -49,6 +53,8 @@ func ExecPodContainerAndGetOutput(ctx context.Context, restConfig *rest.Config,
 	if err != nil {
 		return err
 	}
+	ctx, cancel := context.WithTimeout(ctx, spdyTimeout)
```
Comment (Joibel) [inline workflow/signal/signal.go:56]: Make spdyTimeout a local variable to this function, rather than a global please.

**Review by Joibel — CHANGES_REQUESTED** [review]

File: workflow/signal/signal.go (hunk 1 above)
Comment (fyp711) [reply workflow/signal/signal.go:20]: Thanks for your review, i will fix it. Thanks

File: workflow/signal/signal.go (hunk 2 above)
Comment (fyp711) [reply workflow/signal/signal.go:56]: > Make spdyTimeout a local variable to this function, rather than a global please.

Thanks, i will fix it.

File: workflow/signal/signal.go (hunk 1 above)
Comment (agilgur5) [reply workflow/signal/signal.go:20]: `EXECUTOR_SIGNAL_TIMEOUT` was what I was thinking.

Although I'm not sure that this needs to be configurable in the first place, or as long as 10min.

It isn't entirely clear what we should do after a timeout error either though

Comment (fyp711) [reply workflow/signal/signal.go]: > It isn't entirely clear what we should do after a timeout error either though, requeueing may not fix it

Thanks for your advices,  in our environment, we observe that the pod has actually been removed, but this goroutine will still block here.

Comment (fyp711) [reply workflow/signal/signal.go]: So, in my opinion, we should set a timeout that can never happen. The original purpose of this method change in client-go was also to support timeout cancelable
image

Comment (fyp711) [reply workflow/signal/signal.go]: > So, in my opinion, we should set a timeout that can never happen. The original purpose of this method change in client-go was also to support timeout cancelable

see https://github.com/kubernetes/kubernetes/pull/103177

**Comment by fyp711** [comment]
@Joibel @agilgur5 Hi, could you help me to review again? I have already fixed the issues mentioned earlier. Thanks!

**Comment by Joibel** [comment]
> @Joibel @agilgur5 Hi, could you help me to review again? I have already fixed the issues mentioned earlier. Thanks!

Could you answer my question in #13011 on when you think this might be useful please.

**Comment by fyp711** [comment]
> > @Joibel @agilgur5 Hi, could you help me to review again? I have already fixed the issues mentioned earlier. Thanks!
> 
> Could you answer my question in #13011 on when you think this might be useful please.

Okay, I missed that message. I'm sorry. Now I'll reply

File: docs/environment-variables.md [hunk 3]
```diff
@@ -31,6 +31,7 @@ This document outlines environment variables that can be used to customize behav
...
 | `EVENT_AGGREGATION_WITH_ANNOTATIONS`     | `bool`              | `false`                                                                                     | Whether event annotations will be used when aggregating events.                                                                                                                                                                                                          |
+| `EXECUTOR_SIGNAL_TIMEOUT`                | `time.Duration`     | `10m`                                                                                       | When the executor does not finish after this time, it will times out and simply returns an err                                                                                                                                                                           |
```
Comment (Joibel) [inline docs/environment-variables.md:34]: Suggest rewording the help:
`How long to wait after the executor has been terminated before returning an error`

**Review by Joibel — CHANGES_REQUESTED** [review]

File: docs/environment-variables.md (hunk 3 above)
Comment (fyp711) [reply docs/environment-variables.md:34]: Default I think 10 minutes will do, which is a time that will never happen. It's also a very safe time. Thanks

Comment (fyp711) [reply docs/environment-variables.md:34]: I think that's a good idea, Thanks!

File: docs/environment-variables.md [hunk 4]
```diff
@@ -31,6 +31,7 @@ This document outlines environment variables that can be used to customize behav
...
 | `EVENT_AGGREGATION_WITH_ANNOTATIONS`     | `bool`              | `false`                                                                                     | Whether event annotations will be used when aggregating events.                                                                                                                                                                                                          |
+| `EXECUTOR_SIGNAL_TIMEOUT`                | `time.Duration`     | `10m`                                                                                       | When the executor does not finish after this time, it will times out and simply returns an err. I'm happy with 10 minutes.                                                                                                                                               |
```
Comment (agilgur5) [inline docs/environment-variables.md:34]: looks this was a copy+paste -- that would not be an appropriate documentation line
```suggestion
| `EXECUTOR_SIGNAL_TIMEOUT`                | `time.Duration`     | `10m`                                                                                       | Timeout for a signal to the executor.                                                                                                                                               |
```

What Alan suggested [above](https://github.com/argoproj/argo-workflows/pull/13012#discussion_r1593644855) was:
```suggestion
| `EXECUTOR_SIGNAL_TIMEOUT`                | `time.Duration`     | `10m`                                                                                       | How long to wait after the executor has been terminated before timing out.                                                                                                                                               |
```

I made a small modification to that as well

File: workflow/signal/signal.go (hunk 1 above)
Comment (agilgur5) [reply workflow/signal/signal.go]: > Thanks for your advices, in our environment, we observe that the pod has actually been deleted, but this goroutine will still block here.

Interesting, so a `kubectl exec` is sent to an already terminated Pod and the k8s API Server does not return?
Or the signal via `exec` works but the request does not complete for some reason?

File: docs/environment-variables.md (hunk 3 above)
Comment (agilgur5) [reply docs/environment-variables.md]: Follow-up [below](https://github.com/argoproj/argo-workflows/pull/13012#discussion_r1594592937)

> I'm happy with 10 minutes - although @agilgur5 also had [an opinion](https://github.com/argoproj/argo-workflows/pull/13012#discussion_r1592666100) on that.

But yes, if this is solely to prevent blocking, we can set this lower and don't need it to be configurable. If it were to be configurable it might make sense as a duration _after_ `terminationGracePeriodSeconds`.

I imagine 2min or lower would be safe as a default

Comment (fyp711) [reply docs/environment-variables.md]: @agilgur5 @Joibel In my opinion, i think the best is 2min too. I think it needs to be configurable, it should add some flexibility.

File: docs/environment-variables.md (hunk 4 above)
Comment (fyp711) [reply docs/environment-variables.md:34]: Thanks for helping me revise it.

**Comment by fyp711** [comment]
I updated the code, thanks for help me.

**Comment by fyp711** [comment]
All CI's passed, very successful

File: workflow/signal/signal.go [hunk 5]
```diff
@@ -49,6 +52,9 @@ func ExecPodContainerAndGetOutput(ctx context.Context, restConfig *rest.Config,
 	if err != nil {
 		return err
 	}
+	var executorSignalTimeout = env.LookupEnvDurationOr("EXECUTOR_SIGNAL_TIMEOUT", 2*time.Minute)
```
Comment (Joibel) [inline workflow/signal/signal.go:55]: nit: I'd write this as `executorSignalTimeout := env.LookupEnvDurationOr("EXECUTOR_SIGNAL_TIMEOUT", 2*time.Minute)`, it's more idiomatic golang to do it that way.

**Review by Joibel — CHANGES_REQUESTED** [review]
Thanks, looking good to me, just a minor coding style thing.

I'll let @agilgur5 and you decide on whether this should be configurable. I'm happy with it with the 2minute configurable timeout but don't have strong opinions either way.

File: docs/environment-variables.md (hunk 4 above)
Comment (Joibel) [reply docs/environment-variables.md]: Thanks for fixing this.

File: workflow/signal/signal.go (hunk 5 above)
Comment (fyp711) [reply workflow/signal/signal.go:55]: ok

Comment (fyp711) [reply workflow/signal/signal.go]: fixed

**Review by Joibel — APPROVED** [review]

File: workflow/signal/signal.go (hunk 5 above)
Comment (Joibel) [reply workflow/signal/signal.go]: Thanks. Thank you for your contribution and thank you for your patience with our requests.

Comment (fyp711) [reply workflow/signal/signal.go]: My pleasure. I want the code to be perfect. Thanks.

Comment (fyp711) [reply workflow/signal/signal.go]: All CI's passed

File: workflow/signal/signal.go (hunk 1 above)
Comment (agilgur5) [reply workflow/signal/signal.go]: @fyp711 you didn't answer the above; do you know why it works that way?

File: docs/environment-variables.md (hunk 3 above)
Comment (agilgur5) [reply docs/environment-variables.md]: > Because it is often difficult to update code in production environments, it is relatively easy to update configuration

For sure, but you wrote [above](https://github.com/argoproj/argo-workflows/pull/13012#discussion_r1592743019) that this should never be hit, so in that case, it shouldn't need configuring.

File: workflow/signal/signal.go [hunk 6]
```diff
@@ -49,6 +50,8 @@ func ExecPodContainerAndGetOutput(ctx context.Context, restConfig *rest.Config,
 	if err != nil {
 		return err
 	}
+	ctx, cancel := context.WithTimeout(ctx, 2*time.Minute)
+	defer cancel()
```
Comment (fyp711) [inline workflow/signal/signal.go:54]: @agilgur5 I modified it in this way. Thanks.

Comment (fyp711) [reply workflow/signal/signal.go:54]: I thought about it and it really doesn't need to be configurable.

Comment (fyp711) [reply workflow/signal/signal.go:54]: >  you didn't answer the above; do you know why it works that way?

hi @agilgur5 , Colud i ask which question are you talking about ? That message is too lang, i don't understand which one

Comment (agilgur5) [reply workflow/signal/signal.go:54]: [This one](https://github.com/argoproj/argo-workflows/pull/13012#discussion_r1594598882)

File: workflow/signal/signal.go (hunk 1 above)
Comment (fyp711) [reply workflow/signal/signal.go]: > Interesting, so a `kubectl exec` is sent to an already terminated Pod and the k8s API Server does not return? Or the signal via `exec` works but the request does not complete for some reason?
@agilgur5 I think `exec` has been executed successfully, because that pod is already deleted. I think the reason is apiserver does not return. But actually, I don't know why apiserver didn't return, I tried to analyze it, but it was very difficult. Further k8s community related submissions are this link below. You can have a look. Thank you.
https://github.com/kubernetes/kubernetes/pull/103177

Comment (agilgur5) [reply workflow/signal/signal.go]: Ok, gotcha. Alan and I were trying to understand _why_ this was happening here and in https://github.com/argoproj/argo-workflows/issues/13011#issuecomment-2100244803.

I read the k8s PR, however it similarly doesn't explain _why_. It seems like possibly a k8s API Server bug or `client-go` bug and there should really be an upstream issue for this that we can point to while we use this workaround.

In the interim, I'll leave a comment here noting that it's a workaround and referencing the PR

File: workflow/signal/signal.go [hunk 7]
```diff
@@ -49,6 +50,8 @@ func ExecPodContainerAndGetOutput(ctx context.Context, restConfig *rest.Config,
 	if err != nil {
 		return err
 	}
+	ctx, cancel := context.WithTimeout(ctx, 2*time.Minute)
```
Comment (agilgur5) [inline workflow/signal/signal.go:53]:
```suggestion
    // workaround for when exec does not properly return: https://github.com/kubernetes/kubernetes/pull/103177
	ctx, cancel := context.WithTimeout(ctx, 2*time.Minute)
```

Follow-up to https://github.com/argoproj/argo-workflows/pull/13012#discussion_r1602409330

**Review by agilgur5 — APPROVED** [review]
LGTM. Thanks for reporting this odd bug with k8s and providing the workaround

**Comment by agilgur5** [comment]
I wasn't able to backport this into v3.5.x as it has a merge conflict b/c it depends on #12858, which itself has a merge conflict b/c it depends on #12847 (which itself depends on other things that weren't backported to v3.5.x)

**Comment by fyp711** [comment]
> I wasn't able to backport this into v3.5.x as it has a merge conflict b/c it depends on #12858, which itself has a merge conflict b/c it depends on #12847 (which itself depends on other things that weren't backported to v3.5.x)

Why the v3.5.x branch is different from main branch.

**Comment by fyp711** [comment]
> > I wasn't able to backport this into v3.5.x as it has a merge conflict b/c it depends on #12858, which itself has a merge conflict b/c it depends on #12847 (which itself depends on other things that weren't backported to v3.5.x)
> 
> Why the v3.5.x branch is different from main branch.

Oh, i see you want to add https://github.com/argoproj/argo-workflows/pull/12847 into v3.6.0 Release.

**Comment by agilgur5** [comment]
Correct, the `main` branch tracks _development_, which includes new features and breaking changes of any subsequent minor, and #12847 is a breaking change (as it may not work on older k8s)
